### PR TITLE
Add a PhoneNumber class to handle formatting and actions on numbers

### PR DESCRIPTION
**Pull request by ryanbales**
We value collaboration at Invoca, and focus on peer mentorship as a way for the whole team to learn and thrive. As part of this process, we conduct code reviews during our development cycle. In this phase of the interview, we would like you to review and comment on this Pull Request. 

You can assume that this Pull Request was written by an intern, who is new to professional software development.  Please take about 20-30 minutes to review this Pull Request and comment as if you were reviewing an intern's code.

### CODE REVIEW

File: phone_actions.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class PhoneActions
+
+  def initialize(username, password)
+    @actions_url = "https://#{username}:#{password}@api.<made_up>.com/actions/"
```
Comment (manish-credsimple) [inline phone_actions.rb:4]: The URL/URI  (api.<made_up>.com/actions/) should be a constant.

File: phone_actions.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  end
+
+  def call!(number, message)
+    JSON.parse(RestClient.post("@actions_url/#{text}", {message: message, number: number}))
+  end
+
+  def text!(number, message)
+    JSON.parse(RestClient.post("@actions_url/#{call}", {message: message, number: number}))
+  end
```
Comment (manish-credsimple) [inline phone_actions.rb:13]: call! and text! should just be one method, and pass in the actioon to it.  No reason to have it as two. Also, there's no error handling/exception handling if the post failed for any reason. May be good to have a backoff-retry before logging an error.

File: phone_number.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
+require_relative "./phone_actions"
+
+class PhoneNumber
+  USERNAME = ENV['ACTIONS_USERNAME']
+  PASSWORD = ENV['ACTIONS_PASSWORD']
+
+  def initialize(number)
+    @number = number
```
Comment (manish-credsimple) [inline phone_number.rb:8]: This line's not needed. Just call format_number and pass in the number. It's cleaner this way.

File: phone_number.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+  def call(message)
+    PhoneActions.new(USERNAME, PASSWORD).call!(@number, message)
+  end
+
+  def text_and_call(message)
+    PhoneActions.new(USERNAME, PASSWORD).text!(@number, message)
+    PhoneActions.new(USERNAME, PASSWORD).call!(message, message)
+  end
+
```
Comment (manish-credsimple) [inline phone_number.rb:24]: No need to have 3 methods for text and call. You can just have one method for both and pass in the action.

Also, in text_and_call there's an error for the call method where you need to pass the number to message to for the first argument. 

Error handling/exception logging needs to be here.

File: phone_number.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  def number
+    @number
+  end
```
Comment (manish-credsimple) [inline phone_number.rb:27]: This method is not needed since @number is an instance variable.

File: phone_number.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+  def to_i
+    n = @number.gsub!(/[^0-9]/, "")
+    n.to_i
+  end
```
Comment (manish-credsimple) [inline phone_number.rb:32]: It's not clear why we need a to_i method overridden here. I guess we can have it, but not sure if it's needed in scope.

File: phone_number.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    close_paren = @number.index(')')
+    if open_paren
+      if close_paren
+        @number[(open_paren + 1)..(close_paren - 1)]
+      end
+    else
+      @number.gsub!(/[^0-9]/, "")
+      @number[0..2]
+    end
+  end
```
Comment (manish-credsimple) [inline phone_number.rb:51]: This doesn't take into account here or in the spec if there if parens may be in other places. I.E. -(212-555)-1212 or some other erroneous number put in. Also it should take a number in as the argument instead using the instance variable.

File: phone_number.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+  def to_i
+    n = @number.gsub!(/[^0-9]/, "")
+    n.to_i
+  end
+
+  private
+
+  def format_number
+    "(#{find_area_code}) #{find_exchange}-#{find_line_number}"
```
Comment (manish-credsimple) [inline phone_number.rb:37]: Each of the 3 methods and/or here there needs to be error handling/logging in case number is not properly set. Also this can take in the number as an argument and send that to each of the 3 subsequent methods <find_area_code, find_exchange, find_line_number>

File: phone_number.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+      end
+    else
+      @number.gsub!(/[^0-9]/, "")
+      @number[0..2]
+    end
+  end
+
+  def find_exchange
+    @number.gsub!(/[^0-9]/, "")
+    @number[3..5]
```
Comment (manish-credsimple) [inline phone_number.rb:55]: What if there's a country code in front? I.e. 1-212-555-1212. Also it should take a number in as the argument instead using the instance variable.

File: phone_number.rb [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  def find_line_number
+    @number.gsub!(/[^0-9]/, "")
+    @number[-4..@number.length]
```
Comment (manish-credsimple) [inline phone_number.rb:60]: This woudn't work for a normal number: 3122221212. The -4 will work.

Also, again what if there's a country code in front? IE 1-212-555-1212.

Again it should take a number in as the argument instead using the instance variable.

File: phone_number_spec.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+      phone_number.call("hello world!")
+    end
+  end
+
+  describe "#text_and_call" do
+    it "calls both PhoneActions#text! and PhoneActions#call! methods" do
+      expect_any_instance_of(PhoneActions).to receive(:text!)
+      expect_any_instance_of(PhoneActions).to receive(:call!)
+      phone_number.text_and_call("hello world!")
+    end
```
Comment (manish-credsimple) [inline phone_number_spec.rb:38]: Tests are missing use cases of country codes, invalid entered phone numbers, and making sure that error conditions/exceptions/edge cases are handled properly.
